alphagov/notifications-api
# Auto-switch providers on slow delivery of notifications

This adds a celery task (running every minute) to check for slow delivery of notifications and switch if so.

## How it works

Currently we switch providers automatically when an exception is raised during delivery (either we could not understand the response or they raised an exception). We also want to switch if they are taking a 'long' time to deliver notifications. Based on our functional tests this is deemed to be four minutes. 

The `switch_current_sms_provider_on_slow_delivery` task will run every minute. It will check for any `slow delivery' notifications that occurred in the last ten minutes for the current provider. If more than two occurred, we will switch to the other provider. 

### Avoiding a provider switching loop

Since this will poll in the last ten minutes for any slow delivery notifications, we want to avoid the case where we switching back and forth between providers. Looking in the last ten minutes for slow delivery notifications is not enough, rather we need to look after the point at which we switched to current provider. E.g. If we switched to the current provider on Thursday at 5pm, we look from this point onwards and in the last ten minutes: `thursday 5pm -------> [ten minutes ago ------ current time]`. In this window, we check for slow deliveries.

## How to test

1. Set `FUNCTIONAL_TEST_PROVIDER_SERVICE_ID` and `FUNCTIONAL_TEST_PROVIDER_SMS_TEMPLATE_ID` to the respective ids of your dev service and a sample sms template.

2. Send some notifications with that template and manually override the `updated_at` in the database. This should be set to a date and time more than four minutes after the `sent_at`.

3. Run the celery beat task.

4. Once the above is set, monitor the celery logs to see the task run and the provider switch. It is set to run every minute so you should see it shortly after.  

_Note: This is set to run on production only and is driven by our functional test provider service_
